pkg/tsdb/elasticsearch/response_parser.go

```diff
@@ -56,19 +57,156 @@ func parseResponse(responses []*es.SearchResponse, targets []*Query) (*backend.Q
 
 		queryRes := backend.DataResponse{}
 
-		props := make(map[string]string)
-		err := processBuckets(res.Aggregations, target, &queryRes, props, 0)
-		if err != nil {
-			return &backend.QueryDataResponse{}, err
-		}
-		nameFields(queryRes, target)
-		trimDatapoints(queryRes, target)
+		if isDocumentQuery(target) {
+			err := processDocumentResponse(res, target, timeField, &queryRes)
+			if err != nil {
+				return &backend.QueryDataResponse{}, err
+			}
+			result.Responses[target.RefID] = queryRes
+		} else {
+			props := make(map[string]string)
+			err := processBuckets(res.Aggregations, target, &queryRes, props, 0)
+			if err != nil {
+				return &backend.QueryDataResponse{}, err
+			}
+			nameFields(queryRes, target)
+			trimDatapoints(queryRes, target)
 
-		result.Responses[target.RefID] = queryRes
+			result.Responses[target.RefID] = queryRes
+		}
 	}
 	return &result, nil
 }
 
+func processDocumentResponse(res *es.SearchResponse, target *Query, timeField string, queryRes *backend.DataResponse) error {
+	docs := make([]map[string]interface{}, len(res.Hits.Hits))
+	propNames := make([]string, 0)
+
+	for hitIdx, hit := range res.Hits.Hits {
+		var flattened map[string]interface{}
+		if hit["_source"] != nil {
+			flattened = flatten(hit["_source"].(map[string]interface{}))
+		}
+
+		doc := map[string]interface{}{
+			"_id":       hit["_id"],
+			"_type":     hit["_type"],
+			"_index":    hit["_index"],
+			"sort":      hit["sort"],
+			"highlight": hit["highlight"],
+			"_source":   flattened,
+		}
+
+		for k, v := range flattened {
+			doc[k] = v
+		}
+
+		for key := range doc {
+			// We want to add only unique propNames
+			if isStringInSlice(key, propNames) {
+				continue
+			} else {
+				propNames = append(propNames, key)
+			}
+		}
+
+		docs[hitIdx] = doc
+	}
+
+	size := len(docs)
+	isFilterable := true
+	allFields := make([]*data.Field, len(propNames))
+
+	sortedPropNames := sortPropNames(propNames, timeField)
+
+	for propNameIdx, propName := range sortedPropNames {
+		// Special handling for time field
+		if propName == timeField {
+			timeVector := make([]time.Time, size)
+			for i, doc := range docs {
+				timeString := doc[timeField].(string)
+				timeValue, err := time.Parse(time.RFC3339Nano, timeString)
+				if err != nil {
+					// We skip time values that cannot be parsed
+					continue
+				} else {
+					timeVector[i] = timeValue
+				}
+			}
+			field := data.NewField(timeField, nil, timeVector)
+			field.Config = &data.FieldConfig{Filterable: &isFilterable}
+			allFields[propNameIdx] = field
+			continue
+		}
+
+		propNameValue := findTheFirstNonNilDocValueForPropName(docs, propName)
+		switch propNameValue.(type) {
+		// We are checking for default data types values (float64, int, bool, string)
+		// and default to json.RawMessage if we cannot find any of them
+		case float64:
+			fieldVector := make([]float64, size)
+			for i, doc := range docs {
+				v, _ := doc[propName].(float64)
```

Comment:
this means if the type-conversion fails for the given `doc`'s field, we use it's default-value.

i'm not too happy about this happening for numbers&booleans, but i also don't know anything better 😄 

maybe we could add a comment to these lines to make the decision explicit? not exactly sure what, maybe something like:
```suggestion
				v, _ := doc[propName].(float64) // if the data is invalid, we use the data-type's zero-value
```